src/context/user_context.js

```diff
@@ -18,91 +35,77 @@ export const UserProvider = ({ children }) => {
   }, []);
 
   const handleLogin = async (userCredentials) => {
-    setAuthLoading(true);
+    dispatch({ type: LOGIN_BEGIN });
     try {
       const url = `${rootUrl}/api/v1/auth/login`;
       // We could use fetch or axios. Fetch is native to the browser (doesn't need extra libs), but since we already have axios in use in this repo, I'll stick to using that. It's a little bit cleaner too than raw fetch, but either one is fine.
       const response = await axios.post(url, userCredentials, {
         withCredentials: true,
       });
       console.log(response);
-      setAuthLoading(false);
-      setCurrentUser(response.data.user);
+      dispatch({ type: LOGIN_SUCCESS, payload: response.data.user });
       return true;
     } catch (error) {
       console.error(error);
       const errorPayload =
         error instanceof AxiosError ? error.response.data : error;
-      setAuthLoading(false);
-      setAuthError(errorPayload);
+      dispatch({ type: LOGIN_ERROR, payload: errorPayload });
       return false;
     }
   };
 
   const handleRegister = async (userCredentials) => {
-    setAuthLoading(true);
+    dispatch({ type: REGISTER_BEGIN });
     try {
       const url = `${rootUrl}/api/v1/auth/register`;
       const response = await axios.post(url, userCredentials, {
         withCredentials: true,
       });
       console.log(response);
-      setCurrentUser(response.data.user);
-      setAuthLoading(false);
+      dispatch({ type: REGISTER_SUCCESS, payload: response.data.user });
       return true;
     } catch (error) {
       console.error(error);
       const errorPayload =
         error instanceof AxiosError ? error.response.data : error;
-      setAuthError(errorPayload);
-      setAuthLoading(false);
+      dispatch({ type: REGISTER_ERROR, payload: errorPayload });
       return false;
     }
   };
 
   const handleLogout = async () => {
-    setAuthLoading(true);
+    dispatch({ type: LOGOUT_BEGIN });
     try {
       const url = `${rootUrl}/api/v1/auth/logout`;
       const response = await axios.get(url);
       console.log(response);
-      setCurrentUser(null);
-      setAuthLoading(false);
+      dispatch({ type: LOGOUT_SUCCESS, payload: response.data });
       return true;
     } catch (error) {
       console.error(error);
       const errorPayload =
         error instanceof AxiosError ? error.response.data : error;
-      setAuthError(errorPayload);
-      setAuthLoading(false);
+      dispatch({ type: LOGOUT_ERROR, payload: errorPayload });
       return false;
     }
   };
 
   return (
     <UserContext.Provider
-      value={{
-        handleLogin,
-        handleRegister,
-        handleLogout,
-        currentUser,
-        authLoading,
-        authError,
-      }}
+      value={{ handleLogin, handleRegister, handleLogout, authState }}
```

Comment:
I chose to pass the whole state object `authState`, but if we wanted, we could've kept it as the individual fields: `authLoading: authState.authLoading, authError: authState.authError, currentUser: authState.currentUser` , or even shorter using the spread operator: `...authState`.

Then in the components we could keep using `const { currentUser } = useUserContext();`